src/ripper/ripcddialog.cpp

```diff
@@ -132,36 +128,65 @@ RipCDDialog::RipCDDialog(QWidget* parent)
 
 RipCDDialog::~RipCDDialog() {}
 
-bool RipCDDialog::CheckCDIOIsValid() { return ripper_->CheckCDIOIsValid(); }
-
 void RipCDDialog::closeEvent(QCloseEvent* event) {
   if (working_) {
     event->ignore();
   }
 }
 
 void RipCDDialog::showEvent(QShowEvent* event) {
-  ResetDialog();
-  loader_->LoadSongs();
+  // No need to refresh track list here anymore. On first open this is initiated
+  // by InitializeDevices(). Afterwards track list is always updates when a disc
+  // change is detected due to working through CddaDevice
   if (!working_) {
     ui_->progress_group->hide();
   }
 }
 
-void RipCDDialog::ClickedRipButton() {
-  if (ripper_->MediaChanged()) {
-    QMessageBox cdio_fail(QMessageBox::Critical, tr("Error Ripping CD"),
-                          tr("Media has changed. Reloading"));
-    cdio_fail.exec();
-    ResetDialog();
-    if (CheckCDIOIsValid()) {
-      loader_->LoadSongs();
+void RipCDDialog::InitializeDevices() {
+  Q_ASSERT(!cdda_device_);
+
+  // add all devices to drop down selection
+  for (const DeviceInfo* device_info : cdda_devices_) {
+    ui_->cd_drive_selection->addItem(device_info->friendly_name_);
+  }
+  // ensure that selecting the first device below will emit a
+  // currentIndexChanged signal
+  ui_->cd_drive_selection->setCurrentIndex(-1);
+
+  connect(ui_->cd_drive_selection, SIGNAL(currentIndexChanged(int)),
+          SLOT(DeviceSelected(int)));
+
+  // look for any already connected device, guess that might be the one the user
+  // is interested in and make it the active selection
+  for (int i = 0; i < cdda_devices_.size(); ++i) {
+    DeviceInfo* device_info = cdda_devices_[i];
+    if (device_info->device_) {
+      // found one!
+      ui_->cd_drive_selection->setCurrentIndex(i);
+      return;
     }
-    return;
   }
 
+  // there is no device that is already connected; just select the first one
+  if (!cdda_devices_.isEmpty()) ui_->cd_drive_selection->setCurrentIndex(0);
+}
+
+void RipCDDialog::ClickedRipButton() {
+  Q_ASSERT(cdda_device_);
+
+  // create and connect Ripper instance for this task
+  Ripper* ripper = new Ripper(cdda_device_->raw_cdio(), this);
```

Comment:
What happens to ripper if the media changes now? Does it fail and signal Finished?

Reply:
Yes. Same as before.